DMCompiler/Optimizer/PeepholeOptimizations.cs

```diff
@@ -0,0 +1,607 @@
+        return [
+            DreamProcOpcode.PushNRefs,
+            DreamProcOpcode.CreateList
+        ];
+    }
+
+    public bool CheckPreconditions(List<IAnnotatedBytecode> input, int index) {
+        if (index + 1 >= input.Count) {
+            throw new ArgumentOutOfRangeException("Bytecode index is outside the bounds of the input list.");
+        }
+
+        int pushVal1 = ((((input[index] as AnnotatedBytecodeInstruction)!).GetArgs()[0] as AnnotatedBytecodeInteger)!).Value;
+        int pushVal2 = ((((input[index + 1] as AnnotatedBytecodeInstruction)!).GetArgs()[0] as AnnotatedBytecodeListSize)!).Size;
+
+        return pushVal1 == pushVal2;
+    }
+
+    public void Apply(List<IAnnotatedBytecode> input, int index) {
+        if (index + 1 >= input.Count) {
+            throw new ArgumentOutOfRangeException("Bytecode index is outside the bounds of the input list.");
+        }
+
+        var firstInstruction = Unsafe.As<AnnotatedBytecodeInstruction>(input[index]);
+        int pushVal1 = ((firstInstruction.GetArgs()[0] as AnnotatedBytecodeInteger)!).Value;
+
+        List<IAnnotatedBytecode> args = new List<IAnnotatedBytecode>(1 + pushVal1);
+        args.Add(new AnnotatedBytecodeInteger(pushVal1, new Location()));
+
+        var firstInstructionArgs = firstInstruction.GetArgs();
+        for (int i = 1; i <= pushVal1 && i < firstInstructionArgs.Count; i++) {
+            args.Add(firstInstructionArgs[i]);
+        }
+
+        input.RemoveRange(index, 2);
+        input.Insert(index, new AnnotatedBytecodeInstruction(DreamProcOpcode.CreateListNRefs, 1, args));
+    }
+
+
+}
+
+// Jump [label1]
+// Jump [label2] <- Dead code
+// -> Jump [label1]
+internal sealed class RemoveJumpFollowedByJump : IPeepholeOptimization {
+    public ReadOnlySpan<DreamProcOpcode> GetOpcodes() {
+        return [
+            DreamProcOpcode.Jump,
+            DreamProcOpcode.Jump
+        ];
+    }
+
+    public void Apply(List<IAnnotatedBytecode> input, int index) {
+        input.RemoveAt(index + 1);
+    }
+}
+
+// PushType [type]
+// IsType
+// -> IsTypeDirect [type]
+internal sealed class IsTypeDirect : IPeepholeOptimization {
+    public ReadOnlySpan<DreamProcOpcode> GetOpcodes() {
+        return [
+            DreamProcOpcode.PushType,
+            DreamProcOpcode.IsType
+        ];
+    }
+
+    public void Apply(List<IAnnotatedBytecode> input, int index) {
+        if (index + 1 >= input.Count) {
+            throw new ArgumentOutOfRangeException("Index plus one is outside the bounds of the input list.");
+        }
+
+        var firstInstruction = Unsafe.As<AnnotatedBytecodeInstruction>(input[index]);
+        AnnotatedBytecodeTypeID? pushVal = (firstInstruction.GetArgs()[0] as AnnotatedBytecodeTypeID);
+
+        input.RemoveRange(index, 2);
+        input.Insert(index, new AnnotatedBytecodeInstruction(DreamProcOpcode.IsTypeDirect,
+            new List<IAnnotatedBytecode> { pushVal }));
```

Comment:
`pushVal` is nullable here

Reply:
I'm not sure what's supposed to happen if it's null.